### PR TITLE
Yf fix asm bisulfite error

### LINKED ISSUE #772: Logic of "If statements" for mismatch count is wrong
### Affected tool(s)
CollectAlignmentSummaryMetrics

### Affected version(s)
picard-2.9.0-9-gb9978f5-SNAPSHOT-all
### Description 

The logic of 

`                           
                           if (mismatch && isBisulfiteSequenced &&
                                    record.getReadNegativeStrandFlag() &&
                                    (refBases[refIndex + i] == 'G' || refBases[refIndex + i] == 'g') &&
                                    (readBases[readBaseIndex] == 'A' || readBases[readBaseIndex] == 'a')
                                    || ((!record.getReadNegativeStrandFlag()) &&
                                    (refBases[refIndex + i] == 'C' || refBases[refIndex + i] == 'c') &&
                                    (readBases[readBaseIndex] == 'T') || readBases[readBaseIndex] == 't')) {
                                bisulfiteBase = true;
                                mismatch = false;
                            }`

is different from the original IF statement before [squid:S1066] merge:

`
if (mismatch && isBisulfiteSequenced) {
...}
`

where the condition (mismatch && isBisulfiteSequenced) should be firstly guaranteed. 

#### Expected behavior
 Number of mismatch bases increase 

#### Actual behavior
 Actual number of mismatch bases should be less

### PR DESCRIPTION
**Pull request by yfarjoun**
fixes #772 and adds tests.

three commits are included. 

The first reverts #469 which I take full blame for having reviewed and approved even though it introduced an error. It also adds a test which would have caught the erroneous change.

The second recapitulates the change and should therefore fail the tests.

The third is my proposed change which is cleaner and still passes the tests.

### CODE REVIEW

**Comment by yfarjoun** [comment]
@ktibbett another one for ASM.

**Comment by yfarjoun** [comment]
@jacarey can you suggest a reviewer?

File: src/main/java/picard/analysis/AlignmentSummaryMetricsCollector.java [hunk 1]
```diff
@@ -298,28 +298,20 @@ else if (!record.getReadFailsVendorQualityCheckFlag()) {
...
+
+                            final boolean biSulfiteBase = mismatch && biSulfateMatch;
+                            mismatch = mismatch && !biSulfateMatch;
 
-                            if(mismatch) mismatchCount++;
+                            if (mismatch) mismatchCount++;
 
                             metrics.PF_ALIGNED_BASES++;
-                            if(!bisulfiteBase) nonBisulfiteAlignedBases++;
+                            if (!biSulfiteBase) nonBisulfiteAlignedBases++;
```
Comment (jacarey) [inline src/main/java/picard/analysis/AlignmentSummaryMetricsCollector.java]: bisulfate is a single word and should not be camel caps

File: src/main/java/picard/analysis/AlignmentSummaryMetricsCollector.java [hunk 2]
```diff
@@ -298,28 +298,20 @@ else if (!record.getReadFailsVendorQualityCheckFlag()) {
...
-                                    (refBases[refIndex + i] == 'C' || refBases[refIndex + i] == 'c') &&
-                                    (readBases[readBaseIndex] == 'T') || readBases[readBaseIndex] == 't')) {
-
-                                bisulfiteBase = true;
-                                mismatch = false;
-                            }
+                            boolean mismatch = !SequenceUtil.basesEqual(readBases[readBaseIndex], refBases[refIndex + i]);
+                            final boolean biSulfateMatch = isBisulfiteSequenced && SequenceUtil.bisulfiteBasesEqual(record.getReadNegativeStrandFlag(), readBases[readBaseIndex], refBases[readBaseIndex]);
+
+                            final boolean biSulfiteBase = mismatch && biSulfateMatch;
```
Comment (jacarey) [inline src/main/java/picard/analysis/AlignmentSummaryMetricsCollector.java]: should be isBisulfiteBase

File: src/main/java/picard/analysis/AlignmentSummaryMetricsCollector.java [hunk 3]
```diff
@@ -298,28 +298,20 @@ else if (!record.getReadFailsVendorQualityCheckFlag()) {
...
-                                    (readBases[readBaseIndex] == 'A' || readBases[readBaseIndex] == 'a')
-                                    || ((!record.getReadNegativeStrandFlag()) &&
-                                    (refBases[refIndex + i] == 'C' || refBases[refIndex + i] == 'c') &&
-                                    (readBases[readBaseIndex] == 'T') || readBases[readBaseIndex] == 't')) {
-
-                                bisulfiteBase = true;
-                                mismatch = false;
-                            }
+                            boolean mismatch = !SequenceUtil.basesEqual(readBases[readBaseIndex], refBases[refIndex + i]);
+                            final boolean biSulfateMatch = isBisulfiteSequenced && SequenceUtil.bisulfiteBasesEqual(record.getReadNegativeStrandFlag(), readBases[readBaseIndex], refBases[readBaseIndex]);
```
Comment (jacarey) [inline src/main/java/picard/analysis/AlignmentSummaryMetricsCollector.java]: bisulfite not sulfate

**Comment by jacarey** [comment]
spelling and naming issues.. otherwise 👍

**Pull request merged**